rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1482 @@
+    dnnl_memory_desc_t src_iter_c_desc;
+    /// Weights layer memory descriptor.
+    dnnl_memory_desc_t weights_layer_desc;
+    /// Weights iteration memory descriptor.
+    dnnl_memory_desc_t weights_iter_desc;
+    /// Bias memory descriptor.
+    dnnl_memory_desc_t bias_desc;
+    /// Destination layer memory descriptor.
+    dnnl_memory_desc_t dst_layer_desc;
+    /// Destination iter memory descriptor for hidden state.
+    dnnl_memory_desc_t dst_iter_desc;
+    /// Destination iter memory descriptor for cell state.
+    dnnl_memory_desc_t dst_iter_c_desc;
+    /// Weights peephole memory descriptor.
+    /// This memory descriptor is equal to zero memory descriptor in case of
+    /// non-peephole LSTMs and other non-LSTM RNNs.
+    dnnl_memory_desc_t weights_peephole_desc;
+    /// Weights projection memory descriptor.
+    /// This memory descriptor is equal to zero memory descriptor in case of
+    /// non-projection LSTMs and other non-LSTM RNNs.
+    dnnl_memory_desc_t weights_projection_desc;
+
+    /// Source gradient layer memory descriptor.
+    dnnl_memory_desc_t diff_src_layer_desc;
+    /// Source gradient layer attention memory descriptor.
+    dnnl_memory_desc_t diff_src_layer_attention_desc;
+    /// Source gradient iter memory descriptor for hidden state.
+    dnnl_memory_desc_t diff_src_iter_desc;
+    /// Source gradient iter memory descriptor for cell state.
+    dnnl_memory_desc_t diff_src_iter_c_desc;
+    /// Weights gradient layer memory descriptor.
+    dnnl_memory_desc_t diff_weights_layer_desc;
+    /// Weights gradient iter memory descriptor.
+    dnnl_memory_desc_t diff_weights_iter_desc;
+    /// Bias gradient memory descriptor.
+    dnnl_memory_desc_t diff_bias_desc;
+    /// Destination gradient layer memory descriptor.
+    dnnl_memory_desc_t diff_dst_layer_desc;
+    /// Destination gradient iteration memory descriptor for hidden state.
+    dnnl_memory_desc_t diff_dst_iter_desc;
+    /// Destination gradient iteration memory descriptor for cell state.
+    dnnl_memory_desc_t diff_dst_iter_c_desc;
+    /// Weights gradient peephole memory descriptor.
+    /// This memory descriptor is equal to zero memory descriptor in case of
+    /// non-peephole LSTMs and other non-LSTM RNNs.
+    dnnl_memory_desc_t diff_weights_peephole_desc;
+    /// Weights gradient projection memory descriptor.
+    /// This memory descriptor is equal to zero memory descriptor in case of
+    /// non-projection LSTMs and other non-LSTM RNNs.
+    dnnl_memory_desc_t diff_weights_projection_desc;
+
+    /// RNN cell flags
+    unsigned int flags;
+    /// Activation function used for vanilla_rnn cell kind.
+    /// Must be either #dnnl_eltwise_relu or #dnnl_eltwise_tanh.
+    dnnl_alg_kind_t activation_kind;
+    float alpha;
+    float beta;
+
+} dnnl_rnn_desc_v2_t;
+```
+
+> Note
+>
+> To reduce size of `dnnl_rnn_desc_v2_t` data/diff attention md can be
+> combined into a union with data/diff peephole md.
```

Comment:
Just checking that I'm understanding each option, does this note refer to what @mgouicem has mentioned regarding naming descriptors (`attention_desc` vs. `weights_peephole_desc`)?

Reply:
Partially. In option 1 we reuse `weights_peephole_desc` for `attention_desc` without changing its original name. As a result implementation has places like the following:
```cpp
    memory_desc_t &augru_attention_md() {
        if (with_augru_attention()) return weights_peephole_md_;
        return glob_zero_md;
    }
```
In option 2 we could have separate fields for weights_peephole_desc` and `attention_desc`:
```cpp
    memory_desc_t &augru_attention_md() {
        if (with_augru_attention()) return augru_attention_md_;
        return glob_zero_md;
    }
```
Since option 2 brakes API by introducing new field in `rnn_desc_t` which results in a new `rnn_desc_v2_t` we can also think of other changes like reducing size of `rnn_desc_v2_t` by reusing the same structure fields for different purposes when it is possible. One of such optimizations is to reuse the same filed for `attention_desc`  and `weights_peephole_desc` via union:
```cpp
typedef struct {
...
    union {
        dnnl_memory_desc_t attention_desc;
        dnnl_memory_desc_t weights_peephole_desc;
    } rnn_extension_memory_desc;
...
} dnnl_rnn_desc_v2_t;
// the reason types don't match between the function and the structure is that these are C API and CPP implementation parts.
...
    memory_desc_t &augru_attention_md() {
        if (with_augru_attention()) return rnn_extension_memory_desc.augru_attention_md_;
        return glob_zero_md;
    }
...
```

I used @mgouicem idea of reusing already existing descriptor `weights_peephole_desc` and it is a part of option 1, but there is another way of reusing them in case we choose to break API.

Reply:
Moved it into a separate option C.